SimCalorimetry/HGCalSimProducers/src/HGCHEbackDigitizer.cc

```diff
@@ -11,27 +11,314 @@
+
+  edm::FileInPath fp(fullpath);
+  std::ifstream infile(fp.fullPath());
+  if(!infile.is_open())
+  {
+    throw cms::Exception("FileNotFound") << "Unable to open '" << fullpath << "'" << std::endl;
+  }
+  std::string line;
+  while(getline(infile,line))
+  {
+    int layer;
+    DoseParameters dosePars;
+
+    //space-separated
+    std::stringstream linestream(line);
+    linestream >> layer >> dosePars.a_ >>  dosePars.b_ >> dosePars.c_ >> dosePars.d_ >> dosePars.e_ >> dosePars.f_ >> dosePars.g_ >> dosePars.h_ >> dosePars.i_ >> dosePars.j_;
+
+    result[layer] = dosePars;
+  }
+  return result;
+}
+
+double HGCHEbackSignalScaler::getDoseValue(const int layer, const std::array<double, 8>& radius)
+{
+  double cellDose = std::pow(10, doseMap_[layer].a_ + doseMap_[layer].b_*radius[4] + doseMap_[layer].c_*radius[5] + doseMap_[layer].d_*radius[6] + doseMap_[layer].e_*radius[7]); //dose in grey
+  return cellDose * greyToKrad_; //convert to kRad
+}
+
+double HGCHEbackSignalScaler::getFluenceValue(const int layer, const std::array<double, 8>& radius)
+{
+  double cellFluence = std::pow(10, doseMap_[layer].f_ + doseMap_[layer].g_*radius[0] + doseMap_[layer].h_*radius[1] + doseMap_[layer].i_*radius[2] + doseMap_[layer].j_*radius[3]); //dose in grey
+  return cellFluence;
+}
+
+std::pair<float, float> HGCHEbackSignalScaler::scaleByDose(const HGCScintillatorDetId& cellId,  const std::array<double, 8>& radius)
+{
+  if(doseMap_.empty())
+    return std::make_pair(1., 0.);
+
+  int layer = cellId.layer();
+  double cellDose = getDoseValue(layer, radius); //in kRad
+  constexpr double expofactor = 1./199.6;
+  double scaleFactor = std::exp( -std::pow(cellDose, 0.65) * expofactor );
+
+  double cellFluence = getFluenceValue(layer, radius); //in 1-Mev-equivalent neutrons per cm2
+
+  constexpr double factor = 2. / (2*1e13); //SiPM area = 2mm^2
+  double noise = 2.18 * sqrt(cellFluence * factor);
+
+  if(verbose_)
+  {
+    LogDebug("HGCHEbackSignalScaler") << "HGCHEbackSignalScaler::scaleByDose - Dose, scaleFactor, fluence, noise: "
+                                      << cellDose << " " << scaleFactor << " "
+                                      << cellFluence << " " << noise;
+
+    LogDebug("HGCHEbackSignalScaler") << "HGCHEbackSignalScaler::setDoseMap - layer, a, b, c, d, e, f: "
+                                      << layer << " "
+                                      << doseMap_[layer].a_ << " "
+                                      << doseMap_[layer].b_ << " "
+                                      << doseMap_[layer].c_ << " "
+                                      << doseMap_[layer].d_ << " "
+                                      << doseMap_[layer].e_ << " "
+                                      << doseMap_[layer].f_;
+  }
+
+  return std::make_pair(scaleFactor, noise);
+}
+
+float HGCHEbackSignalScaler::scaleByArea(const HGCScintillatorDetId& cellId, const std::array<double, 8>& radius)
+{
+  float circ = 2 * M_PI * radius[0];
```

Comment:
You can save a couple of cycles if you include ```2 * M_PI``` in the ```constexpr double factor```'s below, and simply compute the edge as ```edge = radius[0] * factor;```